### PR TITLE
Add ByteRangeHelper

### PR DESCRIPTION
**Pull request by jbagga**
Moved [RangeHelper](https://github.com/aspnet/StaticFiles/blob/dev/src/Microsoft.AspNetCore.StaticFiles/Infrastructure/RangeHelpers.cs) from StaticFiles and added tests

See aspnet/Mvc#3702

cc @Tratcher @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    internal static class ByteRangeHelper
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]: Aren't you using this type outside this repo?

File: src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        // 14.35.1 Byte Ranges - If a syntactically valid byte-range-set includes at least one byte-range-spec whose
+        // first-byte-pos is less than the current length of the entity-body, or at least one suffix-byte-range-spec
+        // with a non-zero suffix-length, then the byte-range-set is satisfiable.
+        // Adjusts ranges to be absolute and within bounds.
+        internal static IList<RangeItemHeaderValue> NormalizeRanges(ICollection<RangeItemHeaderValue> ranges, long length)
+        {
+            IList<RangeItemHeaderValue> normalizedRanges = new List<RangeItemHeaderValue>(ranges.Count);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]: `var` all day everyday

**Review by pranavkm — CHANGES_REQUESTED** [review]

**Comment by jbagga** [comment]
@pranavkm @Tratcher Updated

File: src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+{
+    public static class ByteRangeHelper
+    {
+        // 14.35.1 Byte Ranges - If a syntactically valid byte-range-set includes at least one byte-range-spec whose
+        // first-byte-pos is less than the current length of the entity-body, or at least one suffix-byte-range-spec
+        // with a non-zero suffix-length, then the byte-range-set is satisfiable.
+        // Adjusts ranges to be absolute and within bounds.
+        public static IList<RangeItemHeaderValue> NormalizeRanges(ICollection<RangeItemHeaderValue> ranges, long length)
+        {
+            var normalizedRanges = new List<RangeItemHeaderValue>(ranges.Count);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]:
```C#
if (ranges.Count == 0)
{
  return Array.Empty<RangeItemHeaderValue>();
}

var normalizedRanges = new List<RangeItemHeaderValue>(ranges.Count);
...
```

**Review by pranavkm — APPROVED** [review]
Once @Tratcher is happy

File: src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    public static class ByteRangeHelper
+    {
+        // 14.35.1 Byte Ranges - If a syntactically valid byte-range-set includes at least one byte-range-spec whose
+        // first-byte-pos is less than the current length of the entity-body, or at least one suffix-byte-range-spec
+        // with a non-zero suffix-length, then the byte-range-set is satisfiable.
+        // Adjusts ranges to be absolute and within bounds.
+        public static IList<RangeItemHeaderValue> NormalizeRanges(ICollection<RangeItemHeaderValue> ranges, long length)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]: Split out a second method `public static RangeItemHeaderValue NormalizeRange(RangeItemHeaderValue range, long length)`. Most of your tests only apply to this subset of the functionality.

**Comment by jbagga** [comment]
@pranavkm @Tratcher Updated

File: test/Microsoft.AspNetCore.WebUtilities.Tests/ByteRangeHelperTest.cs [hunk 5]
```diff
@@ -2,13 +2,27 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using Microsoft.Net.Http.Headers;
+using System.Collections.Generic;
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.WebUtilities.Tests/ByteRangeHelperTest.cs]: `System.*` goes first

File: src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            if (ranges.Count == 0)
+            {
+                return Array.Empty<RangeItemHeaderValue>();
+            }
+
+            var normalizedRanges = new List<RangeItemHeaderValue>(ranges.Count);
+
+            if (length == 0)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]: Why doesn't this return `Array.Empty<RangeItemHeaderValue>();` too?

File: src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        {
+            var start = range.From;
+            var end = range.To;
+
+            // X-[Y]
+            if (start.HasValue)
+            {
+                if (start.Value >= length)
+                {
+                    // Not satisfiable, skip/discard.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]: Hmm, my idea to make this a sub method just regressed this bit. You removed the continue that would have skipped adding the result. The behavior is now quite different. Is there a test case for this condition?
https://github.com/aspnet/StaticFiles/blob/f73dc5cebea6c08485662f1703d2aba1fc0cd408/src/Microsoft.AspNetCore.StaticFiles/Infrastructure/RangeHelpers.cs#L34

**Review by Tratcher — CHANGES_REQUESTED** [review]

**Comment by jbagga** [comment]
@Tratcher Added test cases to skip invalid ranges. Thank you for pointing it out!

**Review by Tratcher — APPROVED** [review]

**Comment by jbagga** [comment]
@Tratcher @pranavkm Added another method and tests

File: src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+                var bytes = Math.Min(end.Value, length);
+                start = length - bytes;
+                end = start + bytes - 1;
+            }
+
+            var normalizedRange = new RangeItemHeaderValue(start, end);
+            return normalizedRange;
+        }
+
+        public static bool ParseRange(HttpContext context, DateTimeOffset lastModified, EntityTagHeaderValue etag)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]: This is less efficient the the original in StaticFiles. GetTypedHeaders creates a new structure (which you call multiple times) and .Range parses the header again without any caching. If you want to call it ParseRange then at least return the Range object.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/ByteRangeHelper.cs]: Doc comments

File: src/Microsoft.AspNetCore.WebUtilities/Microsoft.AspNetCore.WebUtilities.csproj [hunk 9]
```diff
@@ -13,6 +13,8 @@
 
   <ItemGroup>
     <ProjectReference Include="..\Microsoft.Net.Http.Headers\Microsoft.Net.Http.Headers.csproj" />
+    <ProjectReference Include="..\Microsoft.AspNetCore.Http.Abstractions\Microsoft.AspNetCore.Http.Abstractions.csproj" />
+    <ProjectReference Include="..\Microsoft.AspNetCore.Http.Extensions\Microsoft.AspNetCore.Http.Extensions.csproj" />
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.WebUtilities/Microsoft.AspNetCore.WebUtilities.csproj]: This layering is problematic. WebUtilities should not add these dependencies which makes it impossible to write your method here... I think you'll have to move it to Http.Extensions reduce the inputs.

**Review by Tratcher — CHANGES_REQUESTED** [review]

**Comment by jbagga** [comment]
@Tratcher @pranavkm Added doc comments and moved `ParseRange` to `Extensions`

File: src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    {
+        /// <summary>
+        /// Returns the requested range if the Range Header in the <see cref="HttpContext.Request"/> is valid.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/> associated with the request.</param>
+        /// <param name="requestHeaders">The <see cref="RequestHeaders"/> associated with the given <paramref name="context"/>.</param>
+        /// <param name="lastModified">The <see cref="DateTimeOffset"/> representation of the last modified date of the file.</param>
+        /// <param name="etag">The <see cref="EntityTagHeaderValue"/> provided in the <see cref="HttpContext.Request"/>.</param>
+        /// <returns>A collection of <see cref="RangeItemHeaderValue"/> containing the ranges parsed from the <paramref name="requestHeaders"/>.</returns>
+        public static ICollection<RangeItemHeaderValue> ParseRange(HttpContext context, RequestHeaders requestHeaders, DateTimeOffset lastModified, EntityTagHeaderValue etag)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs]: Some of your MVC file result types won't always have an etag or lastmodified right? What will you do there? or will null work as expected?

**Review by Tratcher — APPROVED** [review]

File: src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+                    if (lastModified > ifRangeHeader.LastModified)
+                    {
+                        ignoreRangeHeader = true;
+                    }
+                }
+                else if (ifRangeHeader.EntityTag != null && !ifRangeHeader.EntityTag.Compare(etag, useStrongComparison: true))
+                {
+                    ignoreRangeHeader = true;
+                }
+                if (ignoreRangeHeader)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs]: new line so this doesn't look like an else-if

**Comment by jbagga** [comment]
@Tratcher I added `null` checks and made `lastModified` and `etag` optional params. Is that acceptable?

File: src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        /// <summary>
+        /// Returns the requested range if the Range Header in the <see cref="HttpContext.Request"/> is valid.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/> associated with the request.</param>
+        /// <param name="requestHeaders">The <see cref="RequestHeaders"/> associated with the given <paramref name="context"/>.</param>
+        /// <param name="lastModified">The <see cref="DateTimeOffset"/> representation of the last modified date of the file.</param>
+        /// <param name="etag">The <see cref="EntityTagHeaderValue"/> provided in the <see cref="HttpContext.Request"/>.</param>
+        /// <returns>A collection of <see cref="RangeItemHeaderValue"/> containing the ranges parsed from the <paramref name="requestHeaders"/>.</returns>
+        public static ICollection<RangeItemHeaderValue> ParseRange(HttpContext context, RequestHeaders requestHeaders, DateTimeOffset? lastModified = null, EntityTagHeaderValue etag = null)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs]: Default params are not allowed in most of our APIs.

File: src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            {
+                // If the validator given in the If-Range header field matches the
+                // current validator for the selected representation of the target
+                // resource, then the server SHOULD process the Range header field as
+                // requested.  If the validator does not match, the server MUST ignore
+                // the Range header field.
+                bool ignoreRangeHeader = false;
+                if (ifRangeHeader.LastModified.HasValue)
+                {
+                    if (lastModified != null && lastModified > ifRangeHeader.LastModified)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http.Extensions/RangeParser.cs]: !lastModified.HasValue

**Review by Tratcher — APPROVED** [review]

**Comment by jbagga** [comment]
Closing as this has been addressed in https://github.com/aspnet/StaticFiles/pull/185

**Pull request closed without merging**